casual/casual-jca/src/main/java/se/laz/casual/jca/pool/NetworkConnectionPool.java

```diff
@@ -0,0 +1,193 @@
+    private ReferenceCountedNetworkConnection getAtIndex(int index)
+    {
+        synchronized (connectionLock)
+        {
+            return connections.get(index);
+        }
+    }
+
+    public NetworkConnection getOrCreateConnection(Address address, ProtocolVersion protocolVersion, NetworkListener networkListener)
+    {
+        if(!this.address.equals(address))
+        {
+            throw new CasualResourceAdapterException("Address mismatch, have: " + this.address + " got: " + address + " for pool with name: " + poolName);
+        }
+        if(disconnected.get())
+        {
+            throw new CasualConnectionException("disconnected");
+        }
+        synchronized (getOrCreateLock)
+        {
+            // create up to pool size # of connections
+            // after that, randomly chose one - later on we can have some better heuristics for choosing which connection to return
+            if (numberOfConnections() == poolSize)
+            {
+                ReferenceCountedNetworkConnection connection = (numberOfConnections() == 1) ? getAtIndex(0) : getAtIndex(getRandomNumber(0, poolSize));
+                connection.increment();
+                connection.addListener(networkListener);
+                return connection;
+            }
+            ReferenceCountedNetworkConnection connection = networkConnectionCreator.createNetworkConnection(address, protocolVersion, networkListener, this, this);
+            addConnection(connection);
+            return connection;
+        }
+    }
+
+    @Override
+    public void closed(ReferenceCountedNetworkConnection networkConnection)
+    {
+        synchronized (getOrCreateLock)
+        {
+            removeConnection(networkConnection);
+        }
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o)
+        {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass())
+        {
+            return false;
+        }
+        NetworkConnectionPool that = (NetworkConnectionPool) o;
+        return Objects.equals(address, that.address) && Objects.equals(poolName, that.poolName);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(address, poolName);
+    }
+
+    @Override
+    public String toString()
+    {
+        return "NetworkConnectionPool{" +
+                "address=" + address +
+                ", connections=" + connections +
+                ", poolName='" + poolName + '\'' +
+                ", poolSize=" + poolSize +
+                ", disconnected=" + disconnected +
+                '}';
+    }
+
+    private static ReferenceCountedNetworkConnection createNetworkConnection(Address address, ProtocolVersion protocolVersion, NetworkListener networkListener, ReferenceCountedNetworkCloseListener referenceCountedNetworkCloseListener, NetworkListener ownListener)
+    {
+        Domain domain = ConfigurationService.getInstance().getConfiguration().getDomain();
```

Comment:
Yeah we create actual network connections in both places.

Reply:
yes - but it is the almost exact same snippet of code  - should we not have it in one place otherwise we might change something for creating with the network pool reference types and forget to change that in the non network pool reference later?